src/transformers/tokenization_utils.py

```diff
@@ -736,6 +745,213 @@ def _tokenize(self, text, **kwargs):
+            return other_lst[: len(lst)] == lst
+
+        def get_comparison_tokens(
+            tokenize_func, text, boundary_token_offset, token_offset, search_length, cache, **tokenize_func_kwargs
+        ):
+            search_text = text[boundary_token_offset : token_offset + search_length]
+            if search_text not in cache:
+                cache[search_text] = tokenize_func(search_text, **tokenize_func_kwargs)
+            return cache[search_text]
+
+        def standardize(text):
+            # Based on ALBERT's preprocessing
+            text = text.replace("``", '"').replace("''", '"')
+            text = unicodedata.normalize("NFKD", text)
+            text = "".join([c for c in text if not unicodedata.combining(c)])
+            text = text.lower()
+            return text
+
+        # Tokenize text
+        tokens = self.tokenize(text, **kwargs)
+
+        # Get maximum search length
+        max_word_length = max([len(word) for word in text.split()])
+        max_space_length = get_max_space_length(text)
+        max_search_length = min(max_word_length + max_space_length, len(text))
+
+        # Initialize token iteration variables
+        boundary_token_index = 0
+        prev_boundary_token_indexes = [0]
+        offsets = [0]
+        i = 0
+        retry = 0
+        while True:
+            token = tokens[i]
+            match_error = False
+
+            # Initialize search variables
+            offset = offsets[i]
+            search_length = 1
+            comparison_tokens = []
+            if retry > 0:
+                # The tokenization from the boundary doesn't match the text, retrying with a previous boundary
+                boundary_token_index = prev_boundary_token_indexes[-retry]
+            elif retry == 0:
+                # Try boundary of the current token
+                boundary_token_index = i
+            else:
+                # instead of failing, tokenize the whole text to get a certain match
+                boundary_token_index = 0
+                search_length = len(text)
+
+            while True:
+                prev_comparison_tokens = comparison_tokens  # for debugging
+                comparison_tokens = get_comparison_tokens(
+                    self.tokenize,
+                    text,
+                    offsets[boundary_token_index],
+                    offset,
+                    search_length,
+                    self._tokenization_cache,
+                    **kwargs
+                )
+
+                target_tokens = tokens[boundary_token_index : i + 1]
+                if is_prefix(target_tokens, comparison_tokens):
+                    # Found a tokenization match
+                    if len(comparison_tokens) > len(target_tokens):
+                        # Handle special cases
+                        relevant_text = standardize(text[offset : offset + search_length])
+                        detokenized = standardize(self._detokenize_for_offsets(token))
```

Comment:
In order to improve the effectiveness of this search, the text should get the same treatment as the tokens (lowercasing, accents removal, etc.), but I'm not sure how it can be done. While it's not perfect, using `standardize` gives great results and for any practical need it is good enough.